Plutarch/Api/V1.hs

```diff
@@ -0,0 +1,626 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Plutarch.Api.V1 (
+  -- * V1 Specific types
+  PScriptContext (..),
+  PTxInfo (..),
+
+  -- * General types, compatible with V1 and V2
+  PScriptPurpose (..),
+
+  -- ** Script
+  PDatum (..),
+  PRedeemer (..),
+  PDatumHash (..),
+  PRedeemerHash (..),
+  PValidatorHash (..),
+  PStakeValidatorHash (..),
+
+  -- ** Value
+  PValue (..),
+  PCurrencySymbol (..),
+  PTokenName (..),
+
+  -- ** Crypto
+  PPubKeyHash (..),
+  PPubKey (..),
+  PSignature (..),
+
+  -- ** Time
+  PPOSIXTime (..),
+  type PPOSIXTimeRange,
+
+  -- ** Interval
+  PInterval (..),
+  PLowerBound (..),
+  PUpperBound (..),
+  PExtended (..),
+  type PClosure,
+
+  -- ** Address
+  PAddress (..),
+  PCredential (..),
+  PStakingCredential (..),
+
+  -- ** Tx
+  PTxInInfo (..),
+  PTxOut (..),
+  PTxOutRef (..),
+  PTxId (..),
+  PDCert (..),
+
+  -- ** AssocMap
+  PMap (..),
+
+  -- ** Others
+  PMaybe (..),
+  PEither (..),
+) where
+
+--------------------------------------------------------------------------------
+
+import Plutarch (PMatch)
+import Plutarch.Lift (PLift (..), PHaskellType, PLiftVia, PBuiltinType)
+import Plutarch.Builtin (PBuiltinList, PIsData, PData, PAsData, PBuiltinPair)
+import Plutarch.DataRepr 
+  (DataReprHandlers (DRHCons, DRHNil), PDataList, PIsDataRepr, PIsDataReprInstances (PIsDataReprInstances), PIsDataReprRepr, pmatchDataRepr, pmatchRepr)
+import Plutarch.Integer (PInteger, PIntegral)
+import Plutarch.ByteString (PByteString)
+import Plutarch.Bool (PBool, PEq, POrd)
+import qualified PlutusTx.Prelude as PlutusTx
+import qualified PlutusTx.AssocMap as AssocMap
+import qualified Plutus.V1.Ledger.Api as Plutus
+import qualified Plutus.V1.Ledger.Crypto as Plutus
+import PlutusTx.Builtins.Internal (BuiltinByteString (..)) -- ctor in-scope for deriving
+import Plutarch.Prelude
+
+--------------------------------------------------------------------------------
+
+---------- V1 Specific types, Incompatible with V2
+
+newtype PTxInfo (s :: k)
+  = PTxInfo
+      ( Term
+          s
+          ( PDataList
+              '[ PBuiltinList PTxInInfo
+               , PBuiltinList PTxOut
+               , PValue
+               , PValue
+               , PBuiltinList PDCert
+               , PBuiltinList (PBuiltinPair PStakingCredential PInteger)
+               , PPOSIXTimeRange
+               , PBuiltinList PPubKeyHash
```

Comment:
Curious question: Shouldn't this be `PBuiltinList (PAsData PPubKeyHash)` - `PPubKeyHash` is not a data value, rather a builtin bytestring - right? So its data representation would be `PAsData PPubKeyHash`? Compared to, say `PTxInfo`, which *is* a data value the entire time - not a builtin that would be converted to a data value in this context.